streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java

```diff
@@ -212,10 +213,11 @@ public void shouldRestoreState() throws Exception {
         );
 
         // we need long enough timeout to by-pass the log manager's InitialTaskDelayMs, which is hard-coded on server side
+        final long waitForPurgeMs = 60000;
         TestUtils.waitForCondition(
-            new RepartitionTopicVerified(currentSize -> currentSize <= PURGE_SEGMENT_BYTES),
-            60000,
-            "Repartition topic " + REPARTITION_TOPIC + " not purged data after 60000 ms."
+            new RepartitionTopicVerified(currentSize -> currentSize <= 4L * PURGE_SEGMENT_BYTES),
```

Comment:
In the builds on Jenkins, this test failed because a future of a delete record call did not complete before all offsets in the repartition topic got committed. When the delete call finally completes there a no commits that trigger a new delete record call. That means, if the size of the repartition topic after the delete call that did not complete before all commits is still larger than `PURGE_SEGMENT_BYTES`, the tests fails.
So increasing the limit in the test makes it more likely that the test succeeds and at the same time it verifies that most of the segments are deleted.
Alternatively, we could decouple deleting records in the repartition topic from committing. We would need to remove the condition `committed > 0` from this code line: 
https://github.com/apache/kafka/blob/56ab2f80343d7d842c1b0e3069833106d4fe3259/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L1117
We do not need to wait for a commit to delete records in the repartition topic, because if the async call to delete records takes a while to complete, there might be offsets to delete independently of the next commit. All commits since the execution of the delete record call might also have committed offsets that still need to be deleted from the repartition topic. Such a decoupling implies an iteration over all tasks of a stream thread even if no offsets were committed, but it does not imply an additional remote call to delete the records since if no records need to be deleted no remote call is executed.
AFAIK, Kafka Streams only guarantees that records are deleted eventually from the repartition topic, so I opted for the change in the test instead of the change in the hot path of the stream thread.
I tested both options and both options seem to work.
I am not sure why this flakiness was surfaced by enabling the state updater, since it could theoretically also happen with disabled state updater. Maybe the additional state updater thread limits the resources available to the embedded Kafka. But also that seems unlikely since the state updater thread should be waiting on a condition variable at that stage in the test.

Reply:
@lucasbru @mjsax @wcarlson5 @aliehsaeedii Could you please sanity check this single change?

Reply:
For my understanding: why can there only be one delete call in-flight at the same time? Where is this prevented?

Reply:
It is prevented here: https://github.com/apache/kafka/blob/7782741262c08e5735f7c8e09727ec37cb5f7f02/streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java#L1782

I do not know why only one delete call should be in-flight at the same time. Maybe to limit the load of admin calls on the brokers?

Reply:
Yeah, that makes sense. Seems to me that the overhead wouldn't be noticable if we execute the code even if we didn't commit anything. And it would make the execution of streams slightly less surprising, so, that's what we are going for right? I vote for purging no matter what.

Reply:
I opened the following PR: https://github.com/apache/kafka/pull/14227

Reply:
Merged https://github.com/apache/kafka/pull/14227